File: lytecache-php/tests/Unit/CountersTest.php
<?php

declare(strict_types=1);

namespace Lytecache\Tests\Unit;

use Lytecache\Exceptions\NotNumericException;
use Lytecache\Tests\TestCase;

final class CountersTest extends TestCase
{
    public function test_incr_starts_missing_keys_at_zero(): void
    {
        $cache = $this->newCache();
        self::assertSame(1, $cache->incr('hits'));
    }

    public function test_incr_accumulates(): void
    {
        $cache = $this->newCache();
        $cache->incr('hits', 5);
        self::assertSame(8, $cache->incr('hits', 3));
    }

    public function test_decr(): void
    {
        $cache = $this->newCache();
        $cache->incr('hits', 10);
        self::assertSame(7, $cache->decr('hits', 3));
    }

    public function test_incr_negative_amount(): void
    {
        $cache = $this->newCache();
        self::assertSame(-5, $cache->incr('hits', -5));
    }

    public function test_incr_float(): void
    {
        $cache = $this->newCache();
        self::assertSame(0.5, $cache->incrFloat('ratio', 0.5));
        self::assertSame(0.75, $cache->incrFloat('ratio', 0.25));
    }

    public function test_incr_float_on_existing_int(): void
    {
        $cache = $this->newCache();
        $cache->incr('n', 5);
        self::assertSame(5.5, $cache->incrFloat('n', 0.5));
    }

    public function test_incr_on_non_numeric_value_throws(): void
    {
        $cache = $this->newCache();
        $cache->set('k', 'not a number');
        $this->expectException(NotNumericException::class);
        $cache->incr('k');
    }

    public function test_incr_on_existing_float_throws(): void
    {
        $cache = $this->newCache();
        $cache->incrFloat('k', 1.5);
        $this->expectException(NotNumericException::class);
        $cache->incr('k');
    }

    public function test_incr_float_on_non_numeric_value_throws(): void
    {
        $cache = $this->newCache();
        $cache->set('k', 'nope');
        $this->expectException(NotNumericException::class);
        $cache->incrFloat('k', 1.0);
    }

    public function test_incr_on_expired_key_starts_over(): void
    {
        $cache = $this->newCache();
        $cache->set('k', 100, ttl: 0.0);
        self::assertSame(1, $cache->incr('k'));
    }
}

File: lytecache-go/counters_test.go
package lytecache_test

import (
	"errors"
	"math"
	"testing"

	lytecache "github.com/lytecache/lytecache-go"
)

func TestIncrStartsMissingKeysAtZero(t *testing.T) {
	c := newTestCache(t)
	n, err := c.Incr("hits", 1)
	if err != nil || n != 1 {
		t.Fatalf("n=%d err=%v", n, err)
	}
}

func TestIncrAccumulates(t *testing.T) {
	c := newTestCache(t)
	if _, err := c.Incr("hits", 5); err != nil {
		t.Fatal(err)
	}
	n, err := c.Incr("hits", 3)
	if err != nil || n != 8 {
		t.Fatalf("n=%d err=%v", n, err)
	}
}

func TestDecr(t *testing.T) {
	c := newTestCache(t)
	if _, err := c.Incr("hits", 10); err != nil {
		t.Fatal(err)
	}
	n, err := c.Decr("hits", 3)
	if err != nil || n != 7 {
		t.Fatalf("n=%d err=%v", n, err)
	}
}

func TestIncrNegativeAmount(t *testing.T) {
	c := newTestCache(t)
	n, err := c.Incr("hits", -5)
	if err != nil || n != -5 {
		t.Fatalf("n=%d err=%v", n, err)
	}
}

func TestIncrFloat(t *testing.T) {
	c := newTestCache(t)
	f, err := c.IncrFloat("ratio", 0.5)
	if err != nil || f != 0.5 {
		t.Fatalf("f=%v err=%v", f, err)
	}
	f, err = c.IncrFloat("ratio", 0.25)
	if err != nil || f != 0.75 {
		t.Fatalf("f=%v err=%v", f, err)
	}
}

func TestIncrFloatOnExistingInt(t *testing.T) {
	c := newTestCache(t)
	if _, err := c.Incr("n", 5); err != nil {
		t.Fatal(err)
	}
	f, err := c.IncrFloat("n", 0.5)
	if err != nil || f != 5.5 {
		t.Fatalf("f=%v err=%v", f, err)
	}
}

func TestIncrOnNonNumericValueErrors(t *testing.T) {
	c := newTestCache(t)
	if err := c.Set("k", "not a number"); err != nil {
		t.Fatal(err)
	}
	_, err := c.Incr("k", 1)
	if !errors.Is(err, lytecache.ErrNotNumeric) {
		t.Fatalf("expected ErrNotNumeric, got %v", err)
	}
}

func TestIncrOnFloatValueErrors(t *testing.T) {
	c := newTestCache(t)
	if _, err := c.IncrFloat("k", 1.5); err != nil {
		t.Fatal(err)
	}
	// Incr (integer-only) on an existing float should error, matching real
	// Redis's INCR-on-a-float-string behavior.
	_, err := c.Incr("k", 1)
	if !errors.Is(err, lytecache.ErrNotNumeric) {
		t.Fatalf("expected ErrNotNumeric, got %v", err)
	}
}

func TestIncrFloatOnNonNumericValueErrors(t *testing.T) {
	c := newTestCache(t)
	if err := c.Set("k", "nope"); err != nil {
		t.Fatal(err)
	}
	_, err := c.IncrFloat("k", 1)
	if !errors.Is(err, lytecache.ErrNotNumeric) {
		t.Fatalf("expected ErrNotNumeric, got %v", err)
	}
}

func TestIncrOnExpiredKeyStartsOver(t *testing.T) {
	c := newTestCache(t)
	if err := c.Set("k", int64(100), lytecache.TTL(0)); err != nil { // expires immediately
		t.Fatal(err)
	}
	n, err := c.Incr("k", 1)
	if err != nil || n != 1 {
		t.Fatalf("expected incr on expired key to start at 1, got n=%d err=%v", n, err)
	}
}

func TestIncrLargeValues(t *testing.T) {
	c := newTestCache(t)
	const big = math.MaxInt64 - 10
	if err := c.Set("k", int64(big)); err != nil {
		t.Fatal(err)
	}
	n, err := c.Incr("k", 5)
	if err != nil || n != big+5 {
		t.Fatalf("n=%d err=%v", n, err)
	}
}
